chat-plugins/mafia.js

```diff
@@ -1,919 +1,916 @@
-		}
-		output += '<button value="/choose vote none" name="send">Abstain</button>';
-
-		if (update) {
-			this.sendRoom('|uhtmlchange|mafia' + this.game.room.gameNumber + 'vote' + this.game.gamestate + this.game.day + '|' + this.game.mafiaWindow(image, output));
+	day(extention, initial) {
+		if (this.phase !== 'night' && !initial) return false;
+		if (this.timer) this.setDeadline('off');
+		this.lynches = {};
+		this.phase = 'day';
+		if (extention && !initial) {
+			// Day stays same
+			this.setDeadline(extention);
 		} else {
-			this.sendRoom('|uhtml|mafia' + this.game.room.gameNumber + 'vote' + this.game.gamestate + this.game.day + '|' + this.game.mafiaWindow(image, output));
+			this.dayNum++;
 		}
+		this.sendRoom(`Day ${this.dayNum}. The hammer count is set at ${this.getHammer()}`, {declare: true});
+		this.updatePlayers();
+		return true;
 	}
 
-	// Targeting mechanics:
-
-	// Targets a single player of side side.
-	singleTarget(side, targetSelf) {
-		this.validTargets = {};
-		for (let i in this.game.players) {
-			let thisSide = this.game.players[i].class.side;
-			if ((side === 'any' || thisSide === side || (side === 'nomafia' && thisSide !== 'mafia')) && (targetSelf || this.game.players[i] !== this)) {
-				this.validTargets[i] = this.game.players[i];
-			}
-		}
+	night(early) {
+		if (this.phase !== 'day') return false;
+		if (this.timer) this.setDeadline('off');
+		this.phase = 'night';
+		this.sendRoom(`Night ${this.dayNum}. PM the host your action, or idle.`, {declare: true});
+		if (!early && this.getPlurality()) this.sendRoom(`Plurality is on ${this.players[this.getPlurality()].name}`);
+		this.updatePlayers();
+		return true;
 	}
 
-	// Triggers after the user has selected their target.
-	onReceiveTarget(target) {
-		if (!this.targeting) {
-			return this.sendRoom("You're not selecting a target right now.");
-		}
-
-		if (target in this.validTargets || target === 'none') {
-			if (target === 'none') {
-				if (this.using) delete this.using;
-				this.toExecute = null;
-			} else {
-				this.target = this.game.players[target];
-			}
-
-			this.updateTarget(this.class.image);
-
-			for (let i in this.game.players) {
-				if (!this.game.players[i].done) {
-					return;
-				}
-			}
-			this.game.progress(true);
+	lynch(user, target) {
+		if (this.phase !== 'day') return false;
+		let player = this.players[user.userid];
+		if (!player && this.dead[user.userid] && this.dead[user.userid].restless) player = this.dead[user.userid];
+		if (!(target in this.players) && target !== 'nolynch') return false;
+		if (player.lynching || (target === player.userid && !this.selfEnabled)) return false;
+		let lynch = this.lynches[target];
+		if (!lynch) {
+			this.lynches[target] = {count: 1, lastLynch: Date.now(), dir: 'up', lynchers: [user.userid]};
+			lynch = this.lynches[target];
 		} else {
-			this.sendRoom("Invalid target");
-		}
+			lynch.count++;
+			lynch.lastLynch = Date.now();
```

Comment:
I forgot to bring this up earlier, im pretty sure that it's possible to get two lynches at the same time through buffering them when the server is lagging (or just getting really lucky with another player). Because JS's sort isnt guaranteed to be stable, you can end up with the wrong plur. IMO, the best way to solve this would be to keep track of how many times people have lynched and just use that for the order instead (first lynch would have count of one, next of 2, etc).

Reply:
This is down to a millisecond of timing, and even a method like that could error due to the same flaw, for example, two lynches at the exact same moment could be placed in the wrong order shifting plurality.

Reply:
Unless I'm vastly misunderstanding something, as long as you assign it and increment it in the same method it wouldn't have that issue. That being said, it's probably unlikely enough to not be an issue in practice.

Reply:
(Talked over everything on discord, decided to leave as-is)